File: firefly-example/src/main/java/com/firefly/example/reactive/coffee/store/service/impl/OrderServiceImpl.java
package com.firefly.example.reactive.coffee.store.service.impl;

import com.firefly.annotation.Component;
import com.firefly.annotation.Inject;
import com.firefly.example.reactive.coffee.store.dao.InventoryDAO;
import com.firefly.example.reactive.coffee.store.dao.OrderDAO;
import com.firefly.example.reactive.coffee.store.dao.ProductDAO;
import com.firefly.example.reactive.coffee.store.model.Order;
import com.firefly.example.reactive.coffee.store.model.Product;
import com.firefly.example.reactive.coffee.store.service.OrderService;
import com.firefly.example.reactive.coffee.store.vo.InventoryOperator;
import com.firefly.example.reactive.coffee.store.vo.InventoryUpdate;
import com.firefly.example.reactive.coffee.store.vo.OrderStatus;
import com.firefly.example.reactive.coffee.store.vo.ProductBuyRequest;
import com.firefly.reactive.adapter.db.ReactiveSQLClient;
import com.firefly.utils.CollectionUtils;
import reactor.core.publisher.Mono;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * @author Pengtao Qiu
 */
@Component
public class OrderServiceImpl implements OrderService {

    @Inject
    private OrderDAO orderDAO;

    @Inject
    private InventoryDAO inventoryDAO;

    @Inject
    private ProductDAO productDAO;

    @Inject
    private ReactiveSQLClient db;


    @Override
    public Mono<Boolean> buy(ProductBuyRequest request) {
        if (request == null) {
            return Mono.error(new IllegalArgumentException("The product request is required"));
        }

        if (request.getUserId() == null || request.getUserId().equals(0L)) {
            return Mono.error(new IllegalArgumentException("The user id is required"));
        }

        if (CollectionUtils.isEmpty(request.getProducts())) {
            return Mono.error(new IllegalArgumentException("The products must bu not empty"));
        }

        return db.newTransaction(c ->
                inventoryDAO.updateBatch(request.getProducts(), InventoryOperator.SUB, c)
                            .doOnSuccess(this::verifyInventory)
                            .then(arr -> productDAO.list(toProductIdList(request), c))
                            .map(products -> toOrders(request, products))
                            .then(orders -> orderDAO.insertBatch(orders, c))
                            .map(orderIdList -> true));
    }

    private void verifyInventory(int[] arr) {
        if (Arrays.stream(arr).anyMatch(i -> i <= 0)) {
            throw new IllegalStateException("The products are not enough");
        }
    }

    private List<Long> toProductIdList(ProductBuyRequest request) {
        return request.getProducts().parallelStream().map(InventoryUpdate::getProductId).collect(Collectors.toList());
    }

    private List<Order> toOrders(ProductBuyRequest request, List<Product> products) {
        return products.parallelStream().map(product -> {
            Order order = new Order();
            order.setUserId(request.getUserId());
            order.setStatus(OrderStatus.FINISHED.getValue());
            order.setAmount(request.getProducts()
                                   .parallelStream()
                                   .filter(i -> i.getProductId().equals(product.getId()))
                                   .map(InventoryUpdate::getAmount)
                                   .findFirst()
                                   .orElseThrow(() -> new IllegalStateException("The product amounts must be more than 0")));
            order.setPrice(product.getPrice());
            order.setTotalPrice(product.getPrice() * product.getAmount());
            order.setProductId(product.getId());
            order.setDescription(product.getDescription());
            return order;
        }).collect(Collectors.toList());
    }
}

File: firefly-example/src/main/kotlin/com/firefly/example/kotlin/coffee/store/service/impl/OrderServiceImpl.kt
package com.firefly.example.kotlin.coffee.store.service.impl

import com.firefly.annotation.Component
import com.firefly.annotation.Inject
import com.firefly.example.kotlin.coffee.store.dao.InventoryDAO
import com.firefly.example.kotlin.coffee.store.dao.OrderDAO
import com.firefly.example.kotlin.coffee.store.dao.ProductDAO
import com.firefly.example.kotlin.coffee.store.model.Order
import com.firefly.example.kotlin.coffee.store.model.Product
import com.firefly.example.kotlin.coffee.store.service.OrderService
import com.firefly.example.kotlin.coffee.store.vo.InventoryOperator
import com.firefly.example.kotlin.coffee.store.vo.OrderStatus
import com.firefly.example.kotlin.coffee.store.vo.ProductBuyRequest
import com.firefly.utils.CollectionUtils
import java.math.BigDecimal
import java.util.*
import java.util.stream.Collectors

/**
 * @author Pengtao Qiu
 */
@Component
class OrderServiceImpl : OrderService {

    @Inject
    private lateinit var orderDAO: OrderDAO

    @Inject
    private lateinit var inventoryDAO: InventoryDAO

    @Inject
    private lateinit var productDAO: ProductDAO

    override suspend fun buy(request: ProductBuyRequest) {
        if (Objects.equals(request.userId, 0L)) {
            throw IllegalArgumentException("The user id is required")
        }

        if (CollectionUtils.isEmpty(request.products)) {
            throw IllegalArgumentException("The products must bu not empty")
        }

        verifyInventory(inventoryDAO.updateBatch(request.products, InventoryOperator.SUB))
        val products = productDAO.list(toProductIdList(request))
        val orders = toOrders(request, products)
        orderDAO.insertBatch(orders)
    }

    private fun verifyInventory(arr: IntArray) {
        if (Arrays.stream(arr).anyMatch { i -> i <= 0 }) {
            throw IllegalStateException("The products are not enough")
        }
    }

    private fun toProductIdList(request: ProductBuyRequest): List<Long> {
        return request.products.parallelStream().map { it.productId }.collect(Collectors.toList<Long>())
    }

    private fun toOrders(request: ProductBuyRequest, products: List<Product>): List<Order> {
        return products.parallelStream().map { product ->
            val amount = getAmount(request, product)
            val totalPrice = BigDecimal.valueOf(product.price).multiply(BigDecimal.valueOf(amount)).toDouble()

            Order(null, OrderStatus.FINISHED.value,
                    amount, product.price, totalPrice,
                    product.id ?: 0,
                    request.userId ?: 0L,
                    product.description,
                    null, null)
        }.collect(Collectors.toList<Order>())
    }

    private fun getAmount(request: ProductBuyRequest, product: Product): Long {
        return request.products.parallelStream()
                .filter { i -> Objects.equals(i.productId, product.id) }
                .map { it.amount }
                .findFirst()
                .orElseThrow({ IllegalStateException("The product amounts must be more than 0") })
    }
}